src/NuGet.Clients/NuGet.PackageManagement.UI/NuGetPackageManagerControlSearchTask.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.UI
+{
+    public sealed class NuGetPackageManagerControlSearchTask : IVsSearchTask
+    {
+        private PackageManagerControl _packageManagerControl;
+        private IVsSearchCallback _searchCallback;
+        private IVsSearchQuery _searchQuery;
+
+        public NuGetPackageManagerControlSearchTask(PackageManagerControl packageManagerControl, uint dwCookie, IVsSearchQuery pSearchQuery, IVsSearchCallback pSearchCallback)
+        {
+            _packageManagerControl = packageManagerControl;
+            _searchCallback = pSearchCallback;
+            _searchQuery = pSearchQuery;
+            Id = dwCookie;
+            ErrorCode = 0;
+            SetStatus(VsSearchTaskStatus.Created);
+
+        }
+        public void Start()
+        {
+            SetStatus(VsSearchTaskStatus.Started);
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                _packageManagerControl.SearchPackagesAndRefreshUpdateCount(_searchQuery.SearchString, true, _searchCallback, this);
+            });
+            SetStatus(VsSearchTaskStatus.Completed);
```

Comment:
I can't remember if RunAsync blocks the rest of the method execution, so if it doesn't you might want to move this into the RunAsync so the search isn't prematurely marked as complete.

Reply:
@Newrad0603 thanks, fixed. RunAsync indeed does not block